cmd/kubeadm/app/cmd/join.go

```diff
@@ -152,206 +153,258 @@ var (
 		`)
 )
 
-// NewCmdJoin returns "kubeadm join" command.
-func NewCmdJoin(out io.Writer) *cobra.Command {
-	cfg := &kubeadmapiv1beta1.JoinConfiguration{}
-	kubeadmscheme.Scheme.Default(cfg)
+// joinOptions defines all the options exposed via flags by kubeadm join.
+// Please note that this structure includes the public kubeadm config API, but only a subset of the options
+// supported by this api will be exposed as a flag.
+type joinOptions struct {
+	cfgPath               string
+	token                 string
+	controlPlane          bool
+	ignorePreflightErrors []string
+	externalcfg           *kubeadmapiv1beta1.JoinConfiguration
+}
 
-	fd := &kubeadmapiv1beta1.FileDiscovery{}
-	btd := &kubeadmapiv1beta1.BootstrapTokenDiscovery{}
+// joinData defines all the runtime information used when running the kubeadm join worklow;
+// this data is shared across all the phases that are included in the workflow.
+type joinData struct {
+	cfg                   *kubeadmapi.JoinConfiguration
+	ignorePreflightErrors sets.String
+	outputWriter          io.Writer
+}
 
-	var token string
-	var cfgPath string
-	var ignorePreflightErrors []string
-	var controlPlane bool
-	var advertiseAddress string
-	var bindPort int32 = kubeadmapiv1beta1.DefaultAPIBindPort
+// NewCmdJoin returns "kubeadm join" command.
+// NB. joinOptions is exposed as parameter for allowing unit testing of
+//     the newJoinData method, that implements all the command options validation logic
+func NewCmdJoin(out io.Writer, joinOptions *joinOptions) *cobra.Command {
+	if joinOptions == nil {
```

Comment:
just to mention: we didn't expose initOptions in the case of `kubeadm init`, since we don't have `init_test.go`. maybe now that we have init phases we can have some proper unit tests for init.